tests/update_workflow_test.go

```diff
@@ -3918,3 +3919,471 @@ func (s *integrationSuite) TestUpdateWorkflow_CompletedSpeculativeWorkflowTask_D
+				CommandType: enumspb.COMMAND_TYPE_SCHEDULE_ACTIVITY_TASK,
+				Attributes: &commandpb.Command_ScheduleActivityTaskCommandAttributes{ScheduleActivityTaskCommandAttributes: &commandpb.ScheduleActivityTaskCommandAttributes{
+					ActivityId:             tv.ActivityID("5"),
+					ActivityType:           tv.ActivityType(),
+					TaskQueue:              tv.TaskQueue(),
+					ScheduleToCloseTimeout: tv.InfiniteTimeout(),
+				}},
+			}}, nil
+		case 2:
+			return nil, nil
+		default:
+			s.Failf("wtHandler called too many times", "wtHandler shouldn't be called %d times", wtHandlerCalls)
+			return nil, nil
+		}
+	}
+
+	atHandler := func(execution *commonpb.WorkflowExecution, activityType *commonpb.ActivityType,
+		activityID string, input *commonpb.Payloads, taskToken []byte) (*commonpb.Payloads, bool, error) {
+		return payloads.EncodeString(tv.String("activity-result")), false, nil
+	}
+
+	poller := &TaskPoller{
+		Engine:              s.engine,
+		Namespace:           s.namespace,
+		TaskQueue:           tv.TaskQueue(),
+		WorkflowTaskHandler: wtHandler,
+		ActivityTaskHandler: atHandler,
+		Logger:              s.Logger,
+		T:                   s.T(),
+	}
+
+	// First WFT, will schedule activity. Also force create a new WFT.
+	_, wt1Resp, err := poller.PollAndProcessWorkflowTaskWithAttemptAndRetryAndForceNewWorkflowTask(false, false, false, false, 1, 1, true, nil)
+	s.NoError(err)
+
+	// Drain 2nd WFT (which is force created as requested) to make all events seem by SDK so following update can be speculative.
+	_, err = poller.HandlePartialWorkflowTask(wt1Resp.GetWorkflowTask(), false)
+	s.NoError(err)
+	s.EqualValues(0, wt1Resp.ResetHistoryEventId)
+
+	// Get shardId so we can unload the shard later
+	ms, err := s.adminClient.DescribeMutableState(testCtx, &adminservice.DescribeMutableStateRequest{
+		Namespace: s.namespace,
+		Execution: tv.WorkflowExecution(),
+	})
+	s.NoError(err)
+	shardId, err := strconv.Atoi(ms.ShardId)
+	s.NoError(err)
+
+	// send update wf request, this will trigger speculative wft
+	go func() {
+		_, _ = s.sendUpdate(tv, "1")
+	}()
+
+	// poll the speculative wft
+	wft1, err1 := s.engine.PollWorkflowTaskQueue(testCtx, &workflowservice.PollWorkflowTaskQueueRequest{
+		Namespace: s.namespace,
+		TaskQueue: tv.TaskQueue(),
+	})
+	s.NoError(err1)
+	s.NotNil(wft1)
+	s.True(len(wft1.TaskToken) > 0) // has valid task token
+	s.True(len(wft1.Messages) > 0)  // has valid message
+	s.Equal(int64(10), wft1.StartedEventId)
+	s.Equal(int64(9), wft1.Messages[0].GetEventId())
+
+	// unload the shard, this will make last speculative wft disappear from server
+	_, err = s.adminClient.CloseShard(testCtx, &adminservice.CloseShardRequest{
+		ShardId: int32(shardId),
+	})
+	s.NoError(err)
+
+	// send another update wf request (with SAME updateId), this will trigger a new speculative wft
+	go func() {
+		_, _ = s.sendUpdate(tv, "1")
+	}()
+	time.Sleep(time.Second) // sleep 1s to make sure update reached to server
```

Comment:
This seems flaky, is there not a better way to ensure that an update reaches the server?
Why is it even required in this test? Can't you immediately poll?

Reply:
I probably don't need it. I can immediately poll.